astrolabe/spec_runner.py

```diff
@@ -52,49 +52,45 @@ def __init__(self, *, client, test_name, cluster_name, specification,
         self.config = configuration
         self.failed = False
 
+        # Initialize attribute used for memoization of connection string.
+        self.__connection_string = None
+
         # Account for platform-specific interrupt signals.
         if sys.platform != 'win32':
             self.sigint = signal.SIGINT
         else:
             self.sigint = signal.CTRL_C_EVENT
 
-        # Validate organization and group.
-        self.get_organization()
-        self.get_group()
-
-    @cached_property
-    def get_organization(self):
-        return get_one_organization_by_name(
+        # Validate and store organization and group.
+        self.organization = get_one_organization_by_name(
             client=self.client,
             organization_name=self.config.organization_name)
-
-    @cached_property
-    def get_group(self):
-        return ensure_project(
+        self.group = ensure_project(
             client=self.client, group_name=self.config.group_name,
-            organization_id=self.get_organization().id)
+            organization_id=self.organization.id)
 
     @property
     def cluster_url(self):
-        return self.client.groups[self.get_group().id].clusters[
+        return self.client.groups[self.group.id].clusters[
             self.cluster_name]
 
-    @cached_property
     def get_connection_string(self):
-        cluster = self.cluster_url.get().data
-        prefix, suffix = cluster.srvAddress.split("//")
-        uri_options = self.spec.maintenancePlan.uriOptions.copy()
-
-        # Boolean options must be converted to lowercase strings.
-        for key, value in uri_options.items():
-            if isinstance(value, bool):
-                uri_options[key] = str(value).lower()
-
-        connection_string = (prefix + "//" + self.config.database_username
-                             + ":" + self.config.database_password + "@"
-                             + suffix + "/?")
-        connection_string += urlencode(uri_options)
-        return connection_string
+        if self.__connection_string is None:
+            cluster = self.cluster_url.get().data
+            prefix, suffix = cluster.srvAddress.split("//")
+            uri_options = self.spec.maintenancePlan.uriOptions.copy()
+
+            # Boolean options must be converted to lowercase strings.
+            for key, value in uri_options.items():
+                if isinstance(value, bool):
+                    uri_options[key] = str(value).lower()
+
+            connection_string = (prefix + "//" + self.config.database_username
+                                 + ":" + self.config.database_password + "@"
```

Comment:
Don't we need to url encode the username and password here?

Reply:
I think you are right. I didn't pay any special attention to it since we will mostly just end up using the default userpass. I will change this in the next commit.